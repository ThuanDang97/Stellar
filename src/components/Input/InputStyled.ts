import styled from 'styled-components'

// Utils
import { pxToRem } from '@utils/theme'

// Components
import { ViewStyled } from '@components/styled-components/ViewStyled'
import { InputStyled } from '@components/styled-components/InputStyled'

// Themes
import { lineStatus } from '@themes/LineStatus'

// Props types
import {
  InputWrapperProps,
  InputLineBottomProps,
} from '@self-types/components/Input.props'

export const InputItemStyled = styled(InputStyled)`
  border: none;
  width: 100%;
  background-color: transparent;
  font-size: ${({ theme }) => pxToRem(theme.typography.fontSize.base)};
  color: ${(props) =>
    props.readOnly ? props.theme.colors.spanishGray : props.theme.colors.shark};
`

export const WrapperInputStyled = styled(ViewStyled)<InputWrapperProps>`
  display: flex;
  width: 100%;
  position: relative;
  border-bottom: ${({ sizeLine, lineColor, theme }) =>
    `${pxToRem(sizeLine)} solid ${lineColor || theme.colors.cello}`};

  &::after {
    content: '';
    width: 15%;
    min-width: ${pxToRem(55)};
    border-top: ${({ theme }) => pxToRem(theme.metrics.borderWidth.md)} solid
      ${({ color, theme }) => color || theme.colors.cello};
    border-right: ${({ theme }) => pxToRem(theme.metrics.borderWidth.md)} solid
      transparent;
    position: absolute;
    top: ${(props) => (props.lineTop ? pxToRem(props.lineTop) : pxToRem(41))};
    display: ${({ display }: { display?: string | false }) =>
      display || 'block'};
  }
`

export const WrapperIconStyled = styled(ViewStyled)`
  width: ${({ theme, width }) =>
    width && typeof width === 'string'
      ? width
      : pxToRem(Number(width) || theme.metrics.dimensions.md)};
  height: ${({ theme, height }) =>
    height ? pxToRem(height) : pxToRem(theme.metrics.dimensions.md)};
  position: relative;
  margin-top: ${(props) => (props.mTop ? pxToRem(props.mTop) : 0)};
  margin-right: ${(props) =>
    props.mRight
      ? pxToRem(props.mRight)
      : pxToRem(props.theme.metrics.dimensions.xs)};
  &.disable-item {
    pointer-events: none;
  }
`

export const PasswordStrength = styled.span`
  display: ${({ display }: { display?: string }) => display || 'none'};
  color: ${({ color, theme }) => color || theme.colors.cello};
  font-size: ${({ theme }) => pxToRem(theme.typography.fontSize.base)};
`

export const GroupLine = styled.div`
  display: flex;
  span {
    border-top: transparent;
  }
`

export const LineBottomInputStyled = styled.span<InputLineBottomProps>`
  width: 40%;
  clip-path: ${(props) => (props.level ? lineStatus[props.level] : '')};
  border-bottom: ${(props) =>
    `${pxToRem(props.theme.metrics.dimensions.xs)} solid ${
      props.borderColor || props.theme.colors.cello
    }`};
  ${(props) => props.mLeft && `margin-left: ${pxToRem(props.mLeft)}`};
  ${(props) => props.mRight && `margin-right: ${pxToRem(props.mRight)}`};
`
