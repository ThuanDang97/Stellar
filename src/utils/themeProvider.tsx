import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import { theme } from '@themes/index'

export const renderWithTheme = (component: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
