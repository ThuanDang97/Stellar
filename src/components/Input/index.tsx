import React, { useMemo } from 'react'
import { useTheme } from 'styled-components'

// Props type
import { InputProps } from '@self-types/components/Input.props'

// Components
import {
  WrapperInputStyled,
  InputItemStyled,
  GroupLine,
  LineBottomInputStyled,
  PasswordStrength,
  WrapperIconStyled,
} from '@components/Input/InputStyled'
import Image from 'next/image'
import { TextStyled } from '@components/styled-components'

const Input = ({
  type,
  placeholder,
  onChange,
  value,
  onBlur,
  statusPassword,
  isStraightLine,
  isLogin,
  iconUrl,
  name,
  error,
  mBottom,
  mTop,
  height,
  pBottom,
  lineTop,
  width,
  mRight,
  readOnly,
  maxLength,
  autocomplete,
  mTopIcon,
}: InputProps) => {
  const theme = useTheme()
  const url = `${iconUrl}${value && '-active'}.svg`
  const isSignUpPassword = type === 'password' && !isLogin && !error

  // Render icon input
  const renderInputIcon = useMemo(() => {
    return (
      iconUrl && (
        <WrapperIconStyled width={width} mRight={mRight} mTop={mTopIcon}>
          <Image src={url} alt="icon" layout="fill" />
        </WrapperIconStyled>
      )
    )
  }, [iconUrl, mRight, mTopIcon, url, width])

  return (
    <>
      <WrapperInputStyled
        lineTop={lineTop}
        height={height}
        pBottom={pBottom}
        mTop={mTop}
        mBottom={mBottom}
        data-testid="inputIcon"
        sizeLine={isSignUpPassword ? 0 : theme.metrics.borderWidth.tiny}
        color={error ? theme.colors.pomegranate : theme.colors.cello}
        lineColor={error ? theme.colors.pomegranate : theme.colors.regentGray}
        display={isSignUpPassword || isStraightLine ? 'none' : 'block'}
      >
        {renderInputIcon}
        {value && name === 'phone' ? (
          <TextStyled
            fontSize={theme.typography.fontSize.base}
            mRight={theme.metrics.dimensions.xs}
          >
            +1
          </TextStyled>
        ) : (
          ''
        )}
        <InputItemStyled
          data-testid="iconInput"
          placeholder={placeholder}
          type={type}
          onChange={onChange}
          value={value}
          name={name}
          onBlur={onBlur}
          readOnly={readOnly}
          maxLength={maxLength}
          autoComplete={autocomplete}
        />
        {isSignUpPassword && (
          <PasswordStrength
            display="block"
            color={
              statusPassword === 'medium'
                ? theme.colors.tangerineYellow
                : statusPassword === 'strong'
                ? theme.colors.shamrock
                : theme.colors.pomegranate
            }
            data-testid="spanCheckPass"
          >
            {statusPassword}
          </PasswordStrength>
        )}
      </WrapperInputStyled>
      {isSignUpPassword && (
        <GroupLine data-testid="WrapperGroupLine">
          <LineBottomInputStyled
            level="weak"
            {...(statusPassword && { borderColor: theme.colors.pomegranate })}
          />
          <LineBottomInputStyled
            level="medium"
            {...(statusPassword && {
              borderColor:
                statusPassword !== 'weak'
                  ? theme.colors.pizzaz
                  : theme.colors.silver,
            })}
            mLeft={-3}
            mRight={-3}
          />
          <LineBottomInputStyled
            level="strong"
            {...(statusPassword && {
              borderColor:
                statusPassword === 'strong'
                  ? theme.colors.caribbeanGreen
                  : theme.colors.silver,
            })}
          />
        </GroupLine>
      )}
    </>
  )
}

export default React.memo(Input)
