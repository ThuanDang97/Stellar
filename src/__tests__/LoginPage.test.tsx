import React from 'react'
import { cleanup, fireEvent, screen } from '@testing-library/react'

// Utils
import { renderWithTheme } from '@utils/themeProvider'

// Pages
import LoginPage from '@pages/login'
import Input from '@components/Input'

describe('LoginPage', () => {
  beforeEach(() => {
    cleanup()
    renderWithTheme(<LoginPage />)
  })

  it('should render loginForm', () => {
    const loginForm = screen.getByTestId('loginForm')

    expect(loginForm).toBeInTheDocument()
  })

  it('matches snapshot', () => {
    const { asFragment } = renderWithTheme(<LoginPage />)

    expect(asFragment()).toMatchSnapshot()
  })

  const onChangeValue = jest.fn()
  const setup = () => {
    const utils = renderWithTheme(<LoginPage />)
    const input = screen.getAllByTestId('iconInput') as HTMLInputElement[]
    return {
      input,
      ...utils,
    }
  }
  test('Should onChange value when input value', async () => {
    renderWithTheme(<Input onChange={onChangeValue} />)
    const { input } = setup()
    fireEvent.change(input[0], { target: { value: 'Cheese pocket' } })
    expect(input[0].value).toBe('Cheese pocket')
  })
})
