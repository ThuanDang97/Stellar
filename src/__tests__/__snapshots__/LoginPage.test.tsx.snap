// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ViewStyled-sc-rqe5sa-0 ViewStyled__WrapperContent-sc-rqe5sa-6 fTsvHV gVWyht"
  >
    <div
      class="ViewStyled-sc-rqe5sa-0 hSHZsk"
      height="60"
      width="250"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="stellar-logo"
          data-nimg="fill"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="ViewStyled-sc-rqe5sa-0 kLQFtm"
      display="flex"
    >
      <div
        class="ViewStyled-sc-rqe5sa-0 hxHJAe"
        data-testid="img"
        height="450"
        width="inherit"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="Jumping man image"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
          />
          <noscript />
        </span>
      </div>
      <div
        class="ViewStyled-sc-rqe5sa-0 bEaHGp"
        display="flex"
        width="inherit"
      >
        <h1
          class="TextStyled-sc-9xia41-0 TitleStyled-sc-t58pjn-0 iPgIes WstJK"
          color="#000"
          font-family="AdobeCleanExtraBold"
        >
          Login
        </h1>
        <form
          class="ViewStyled-sc-rqe5sa-0 ViewStyled__BoxShadowStyled-sc-rqe5sa-2 kjqdjo eFQizc"
          data-testid="loginForm"
          display="flex"
          novalidate=""
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperInputStyled-sc-9gj86m-1 UsSdI brcjID"
            color="#203F59"
            data-testid="inputIcon"
            display="block"
          >
            <div
              class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperIconStyled-sc-9gj86m-2 fTsvHV kHphsF"
            >
              <span
                style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
              >
                <img
                  alt="icon"
                  data-nimg="fill"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </span>
            </div>
            <input
              class="InputStyled-sc-1k09cqb-0 InputStyled__InputItemStyled-sc-9gj86m-0 hnxmhM NDCRX"
              data-testid="iconInput"
              name="email"
              placeholder="Enter your email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperInputStyled-sc-9gj86m-1 bNeJX brcjID"
            color="#203F59"
            data-testid="inputIcon"
            display="block"
          >
            <div
              class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperIconStyled-sc-9gj86m-2 fTsvHV kHphsF"
            >
              <span
                style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
              >
                <img
                  alt="icon"
                  data-nimg="fill"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </span>
            </div>
            <input
              class="InputStyled-sc-1k09cqb-0 InputStyled__InputItemStyled-sc-9gj86m-0 hnxmhM NDCRX"
              data-testid="iconInput"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <button
            class="ButtonStyled-sc-1ys9v52-0 iAfDLq"
            data-testid="button"
            disabled=""
            type="submit"
          >
            Login
          </button>
          <a
            class="LinkStyled__ButtonLinkStyled-sc-626068-1 hvsWbe"
            data-testid="link"
            href="/forgot-password"
          >
            Forgot Password?
          </a>
        </form>
        <p
          class="TextStyled-sc-9xia41-0 jjVoJj"
        >
          <span
            class="TextStyled-sc-9xia41-0 loORFZ"
            color="#676767"
          >
            Donâ€™t have an account?
          </span>
          <a
            class="LinkStyled__ButtonLinkStyled-sc-626068-1 lfTGbL"
            data-testid="link"
            display="inline"
            href="/sign-up"
          >
            Signup Now
          </a>
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
